pkg/util/purge/purge.go

```diff
@@ -0,0 +1,323 @@
+			rc.log.Infof("deleting private endpoint connection %s/%s/%s", *resourceGroup.Name, *pls.Name, *peconn.Name)
+			if !rc.dryRun {
+				_, err := rc.privatelinkservicescli.DeletePrivateEndpointConnection(ctx, *resourceGroup.Name, *pls.Name, *peconn.Name)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// CleanAAD does the cleaning of the Active Directory
+//     - clean unsused app registration
+//     - clean unsused role assignments
+func (rc *ResourceCleaner) CleanAAD(ctx context.Context) error {
+	err := rc.cleanApps(ctx)
+	if err != nil {
+		return err
+	}
+
+	return rc.cleanRoleAssignments(ctx)
+
+}
+
+// cleanRoleAssignments lists role assignments from the subscription and clear unused
+// the deletion work by:
+//     - checking whether the coresponding resource group exists (RS and RG have the same name)
+//     - check whether the assigned service principal exists
+//     - the role assignment belongs to one of RG: v4-westeurope, v4-eastus, v4-australiasoutheast
+func (rc *ResourceCleaner) cleanRoleAssignments(ctx context.Context) error {
+	roleAssignments, err := rc.roleassignmentcli.List(ctx, "")
+	if err != nil {
+		return err
+	}
+
+	// iterate over all roleAssignments and delete everything not associated with keep cluster
+	for roleAssignments.NotDone() {
+		pageResults := roleAssignments.Values()
+
+		for _, ra := range pageResults {
+			_, err := rc.resourcegroupscli.Get(ctx, *ra.Name)
+			if err != nil {
+				if detailedErr, ok := err.(autorest.DetailedError); ok &&
+					detailedErr.StatusCode == http.StatusNotFound {
+					continue
+				}
+				return err
+			}
+
+			_, err = rc.serviceprincipalcli.Get(ctx, *ra.PrincipalID)
+			if err != nil {
+				if detailedErr, ok := err.(autorest.DetailedError); ok &&
+					detailedErr.StatusCode == http.StatusNotFound {
+					continue
+				}
+				return err
+			}
+
+			// skip if the role assignment is not in the defined resource group
+			match := false
+			for _, region := range rc.resourceRegions {
+				if strings.HasSuffix(*ra.Scope, fmt.Sprintf("%s/providers/Microsoft.Network/virtualNetworks/dev-vnet", region)) {
+					match = true
+				}
+			}
+			if !match {
+				continue
+			}
+
+			if !rc.dryRun {
+				_, err := rc.roleassignmentcli.DeleteByID(ctx, *ra.ID)
+				if err != nil {
+					return err
+				}
+			}
+			rc.log.Infof("Deleting role assignment: %s", *ra.Name)
+		}
+
+		err := roleAssignments.NextWithContext(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// CleanApps lists all apps from the subscription and clear unused
+// the deletion works by:
+//     - checking whether coresponding resource group exists (the name of app and resource group matches)
+//     - if there is no existing resource group and the application url has prefix: 'https://az.aro.azure.com/' delete
+//     - delete coresponding service principal
+func (rc *ResourceCleaner) cleanApps(ctx context.Context) error {
```

Comment:
you need to consider what it will take to make you comfortable to run this on the Microsoft corporate AAD, if that is even feasible.